cmake/cmaize/project/project.cmake

```
@@ -0,0 +1,145 @@
+    #      the ``VERSION <version>`` keyword for CMake ``project()``
+    #      (`link <cmake_project_>`__), and the value is passed to
+    #      a CMake ``project()`` call if the project does not exist yet.
+    #    * **DESCRIPTION** (*desc*) --
+    #      Description of the project. Parallels the 
+    #      ``DESCRIPTION <project-description-string>`` keyword for 
+    #      CMake ``project()`` (`link <cmake_project_>`__), and the value
+    #      is passed to a CMake ``project()`` call if the project does not
+    #      exist yet.
+    #    * **HOMEPAGE_URL** (*desc*) --
+    #      Homepage URL for the project. Parallels the
+    #      ``HOMEPAGE_URL <url-string>`` keyword for CMake ``project()``
+    #      (`link <cmake_project_>`__), and the value is passed to
+    #      a CMake ``project()`` call if the project does not exist yet.
+    #    * **LANGUAGES** (*list[desc]*) --
+    #      Languages supported by the project. These languages are passed to
+    #      the LANGUAGES keyword of the CMake ``project()`` call if the project
+    #      does not exist yet.
+    #
+    # :returns: ``self`` will be set to the newly constructed
+    #           ``CMaizeProject`` object.
+    # :rtype: CMaizeProject
+    # 
+    # .. Reference definitions
+    # .. _cmake_project: https://cmake.org/cmake/help/latest/command/project.html
+    #]]
+    cpp_constructor(CTOR CMaizeProject str args)
+    function("${CTOR}" self _ctor_name)
+
+        set(_ctor_lists LANGUAGES)
+        cmake_parse_arguments(_ctor "" "" "${_ctor_lists}" ${ARGN})
+
+        CMaizeProject(SET "${self}" name "${_ctor_name}")
+
+        CMaizeProject(SET "${self}" languages "${_ctor_LANGUAGES}")
+
+        # If the project name is different than the current project, assume
+        # a new CMake project needs to be created
+        if(NOT "${_ctor_name}" STREQUAL "${PROJECT_NAME}")
+            project("${_ctor_name}" ${ARGN})
+        endif()
```

Maybe throw an error instead?

---

Ah, I didn't realize calling `project` was going to be required. This can definitely be an error instead.